server/src/main/java/org/elasticsearch/search/aggregations/AggregatorsReducer.java

```diff
@@ -19,24 +19,28 @@
  */
 public final class AggregatorsReducer implements Releasable {
 
-    private final Map<String, AggregatorReducer> aggByName = new HashMap<>();
-    private final AggregationReduceContext context;
-    private final int size;
+    private final Map<String, AggregatorReducer> aggByName;
 
-    public AggregatorsReducer(AggregationReduceContext context, int size) {
-        this.context = context;
-        this.size = size;
+    /**
+     * Solo constructor
+     *
+     * @param proto The prototype {@link InternalAggregations} we are aggregating.
+     * @param context The aggregation context
+     * @param size The number of {@link InternalAggregations} we are aggregating.
+     */
+    public AggregatorsReducer(InternalAggregations proto, AggregationReduceContext context, int size) {
+        aggByName = new HashMap<>(proto.asList().size());
+        for (InternalAggregation aggregation : proto) {
+            aggByName.put(aggregation.getName(), aggregation.getReducer(context.forAgg(aggregation.getName()), size));
+        }
```

Comment:
I think it be assumed thatmwe can reconstruct all the aggregations by using  one of the InternalAggregations we are going to reduce. They all contain all the sub-aggregations.

Reply:
So I cannot assume this, therefore the PR only sizes the map because of rollups.  Still sizing the map is good win.